pkg/app/piped/livestatereporter/cloudrun/report.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2022 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudrun
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+
+	"github.com/pipe-cd/pipecd/pkg/app/piped/livestatestore/cloudrun"
+	"github.com/pipe-cd/pipecd/pkg/app/server/service/pipedservice"
+	"github.com/pipe-cd/pipecd/pkg/config"
+	"github.com/pipe-cd/pipecd/pkg/model"
+)
+
+type applicationLister interface {
+	ListByCloudProvider(name string) []*model.Application
+}
+
+type apiClient interface {
+	ReportApplicationLiveState(ctx context.Context, req *pipedservice.ReportApplicationLiveStateRequest, opts ...grpc.CallOption) (*pipedservice.ReportApplicationLiveStateResponse, error)
+	ReportApplicationLiveStateEvents(ctx context.Context, req *pipedservice.ReportApplicationLiveStateEventsRequest, opts ...grpc.CallOption) (*pipedservice.ReportApplicationLiveStateEventsResponse, error)
+}
+
+type Reporter interface {
+	Run(ctx context.Context) error
+	ProviderName() string
+}
+
+type reporter struct {
+	provider              config.PipedCloudProvider
+	appLister             applicationLister
+	stateGetter           cloudrun.Getter
+	apiClient             apiClient
+	snapshotFlushInterval time.Duration
+	logger                *zap.Logger
+
+	snapshotVersions map[string]model.ApplicationLiveStateVersion
+}
+
+func NewReporter(cp config.PipedCloudProvider, appLister applicationLister, stateGetter cloudrun.Getter, apiClient apiClient, logger *zap.Logger) Reporter {
+	logger = logger.Named("cloudrun-reporter").With(
+		zap.String("cloud-provider", cp.Name),
+	)
+	return &reporter{
+		provider:              cp,
+		appLister:             appLister,
+		stateGetter:           stateGetter,
+		apiClient:             apiClient,
+		snapshotFlushInterval: 10 * time.Minute,
+		logger:                logger,
+		snapshotVersions:      make(map[string]model.ApplicationLiveStateVersion),
+	}
+}
+
+func (r *reporter) Run(ctx context.Context) error {
+	r.logger.Info("start running app live state reporter")
+
+	r.logger.Info("waiting for livestatestore to be ready")
+	if err := r.stateGetter.WaitForReady(ctx, 10*time.Minute); err != nil {
+		r.logger.Error("livestatestore was unable to be ready in time", zap.Error(err))
+		return err
+	}
+
+	snapshotTicker := time.NewTicker(r.snapshotFlushInterval)
+	defer snapshotTicker.Stop()
+
+	for {
+		select {
+		case <-snapshotTicker.C:
```

Comment:
This means that the state will be updated every 10 minutes, right?
Do we have any way to make it refresh faster on UI?

Reply:
@nghialv 

> This means that the state will be updated every 10 minutes, right?

Exactly.

> Do we have any way to make it refresh faster on UI?

Currently, No we don't. Since livestatestore fetch live state at 15s intervals and driftdetection and livestatereporter refer to it. But as one idea, it may be better that adding `forceRefresh` flag into `GetApplicationLiveStateRequest` in order to get latest live state synchronously.
BTW, It might be a good idea to make this interval one minute like as driftdetection.